_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,590 @@
+
+                    consumer.acknowledge(record, AcknowledgeType.REJECT); // #4
+                    break; // #5
+                } catch (Exception e ) {
+                    System.out.println("Failed to process the record- Offset: " + record.offset() + " Value: " + record.value());
+
+                    consumer.acknowledge(record, AcknowledgeType.RELEASE); // #6
+                }
+                consumer.acknowledge(record, AcknowledgeType.ACCEPT); // #7
+            }
+            consumer.commitSync(); // #8
+        }
+    }
+
+    private static void processRecord(ConsumerRecord<String, String> record) throws JsonProcessingException {
+        JsonNode json = OBJECT_MAPPER.readTree(record.value());
+        System.out.println("Processed record with offset " + record.offset() + ": " + json.toPrettyString());
+    }
+
+    private static void handlePoisonRecord(ConsumerRecord<String, String> record, KafkaProducer<String, String> producer) {
+        producer.send(new ProducerRecord<>("dead-letter-queue-topic", record.key(), record.value()));
+    }
+}
+
+```
+#1. In this example, the consumer immediately releases a record if its key is set to "skip", just to demonstrate the retry of delivery attempts. The record is fetched again in the next call to <code>poll()</code> and eventually hits the maximum delivery attempts and is transitioned into <b>Archived</b> state because the client never processes it. Once archived, this record is no longer be available for another delivery. 
+
+#2. If record's key is "skip", it exits the loop for processing each record.
+
+#3. Also when a record value cannot be mapped to a valid JSON object, the processing method throws a <code>JsonProcessingException</code> and the client handles it as a poison record. Since dead letter queue is not supported yet, users have to implement this themselves, similar to this example. 
+
+#4. Once a poison record is sent to the dead letter queue topic, the consumer client rejects it so that it gets archived immediately. 
+
+#5. After rejecting a bad record, it exits the loop for processing each record.
+
+#6. If the client encountered an error not caused by the JSON mapping, it releases the record for another attempt because it could be a transient failure.
+
+#7. When no exception occurred during the processing, the consumer client accepts the record. Once all the records in the batch are processed and acknowledged individually, these states are stored locally in the consumer. 
+
+#8. Then the client has to call <code>commitSync()</code> or <code>commitAsync()</code> to commit the state to the internal topic, <code>__share_group_state</code>. In #2 and #3, it exits the loop for processing each record, and commits the state by calling <code>commitSync()</code>. The records in the batch that were not processed or acknowledged yet, will be presented to the consumer client again as part of the same acquisition, therefore their delivery count will not be incremented.
```

Comment:
Genuine question as I don't know the answer, can you use auto commit with a share group? or do you have to explicitly commit yourself all the time?

Reply:
Then it's probably something worth highlighting